integration/read_write_mode_test.go

```diff
@@ -76,3 +55,106 @@ func TestReadWriteMode(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, []string{"__name__", "foo"}, labelNames)
 }
+
+func TestReadWriteModeRecordingRule(t *testing.T) {
+	s, err := e2e.NewScenario(networkName)
+	require.NoError(t, err)
+	defer s.Close()
+
+	c, backendInstance := startReadWriteModeCluster(
+		t,
+		s,
+		map[string]string{
+			// Evaluate rules often and with no delay, so that we don't need to wait for metrics to show up.
+			"-ruler.evaluation-interval":       "2s",
+			"-ruler.poll-interval":             "2s",
+			"-ruler.evaluation-delay-duration": "0",
+		},
+	)
+
+	// Create a recording rule
+	record := yaml.Node{}
+	testRuleName := "test_rule"
+	record.SetString(testRuleName)
+
+	expr := yaml.Node{}
+	expr.SetString("sum(test_series)")
+
+	ruleGroup := rulefmt.RuleGroup{
+		Name:     "test_rule_group",
+		Interval: 1,
+		Rules: []rulefmt.RuleNode{
+			{
+				Record: record,
+				Expr:   expr,
+			},
+		},
+	}
+
+	require.NoError(t, c.SetRuleGroup(ruleGroup, "test_rule_group_namespace"))
+
+	require.NoError(t, backendInstance.WaitSumMetricsWithOptions(e2e.GreaterOrEqual(1), []string{"cortex_prometheus_rule_evaluations_total"}, e2e.WaitMissingMetrics))
+	ruleEvaluationsBeforePush, err := backendInstance.SumMetrics([]string{"cortex_prometheus_rule_evaluations_total"})
+	require.NoError(t, err)
+
+	// Push data that should be captured by the recording rule
+	pushTime := time.Now()
+	series, _, _ := generateSeries("test_series", pushTime, prompb.Label{Name: "foo", Value: "bar"})
+
+	res, err := c.Push(series)
+	require.NoError(t, err)
+	require.Equal(t, 200, res.StatusCode)
+
+	// Wait for recording rule to evaluate
+	require.NoError(t, backendInstance.WaitSumMetrics(e2e.Greater(ruleEvaluationsBeforePush[0]), "cortex_prometheus_rule_evaluations_total"))
```

Comment:
Rule could have evaluated again just before push. Perhaps we can wait for two subsequent evaluations here?

Reply:
Why don't we do the opposite? First we push the data, then we configure the recording rule. Mimir guarantees the pushed data will be queried by the first evaluation of the recording rule then.

Reply:
Yes, that's a good idea.

Reply:
Good idea, changed in 8368636.